tsdb/head.go

```diff
@@ -424,6 +438,31 @@ func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOpti
 
 func mmappedChunksDir(dir string) string { return filepath.Join(dir, "chunks_head") }
 
+func (h *Head) ApplyConfig(cfg *config.Config) error {
+	if h.opts.EnableExemplarStorage && cfg.StorageConfig.ExemplarsConfig.MaxExemplars > -1 {
+		// ugly hack to get the previous default size
+		if cfg.StorageConfig.ExemplarsConfig.MaxExemplars == 0 {
+			cfg.StorageConfig.ExemplarsConfig.MaxExemplars = 100000
+		}
+		switch h.exemplars.(type) {
+		case noopExemplarStorage:
+			e, err := NewCircularExemplarStorage(cfg.StorageConfig.ExemplarsConfig.MaxExemplars, h.reg)
+			if err != nil {
+				return err
+			}
+			h.exemplars = e
+			// Head uses opts.MaxExemplars in combination with opts.EnableExemplarStorage
+			// to decide if it should pass exemplars along to it's exemplar storage, so we
+			// need to update opts.MaxExemplars here.
```

Comment:
I think we should only use `opts.EnableExemplarStorage` for deciding and not `opts.MaxExemplars`
Reasons:
* `opts` is a config given for starting Head and would be error prone to keep it up to date with dynamic fields like `opts.MaxExemplars`. It is best to get these dynamic config from the exemplar storage itself.
* There could be a case where `opts.MaxExemplars` is made >0 from <0 while the appender has not committed, in which case we should prolly not have discarded those exemplars and have kept in appender and committed.

Reply:
There's a few additional changes that we'd have to make in order to do this.

> opts is a config given for starting Head and would be error prone to keep it up to date with dynamic fields like opts.MaxExemplars. It is best to get these dynamic config from the exemplar storage itself.

All config for the exemplar storage is passed through TSDB head already, and opts was what we were using before. I was trying to make this change as small as possible. If opts has only the `EnableExemplarStorage` bool and not the `MaxExemplars` int then we need `ValidateExemplar` to return a new error type if `MaxExemplars` is `< 0`.

> There could be a case where opts.MaxExemplars is made >0 from <0 while the appender has not committed, in which case we should prolly not have discarded those exemplars and have kept in appender and committed.

I'm not sure what you're saying here/suggesting we change?

Reply:
Ignore my above comment. I was thinking about interaction between and open appender and resize and what should be the right way to append exemplars in that case. So my new question would be, is it fine to only ingest partial exemplars from a scrape? (although rare case, but can happen during a config reload)

Reply:
> So my new question would be, is it fine to only ingest partial exemplars from a scrape? (although rare case, but can happen during a config reload)

I think for now this would be okay, but we could explore improvements in the future.